src/buffer.cc

```diff
@@ -204,6 +205,85 @@ Result Buffer::CompareRMSE(Buffer* buffer, float tolerance) const {
+                                                    uint32_t num_bins) const {
+  assert(num_bins == 256);
+  std::vector<int64_t> bins(num_bins, 0.0);
+  auto* buf_ptr = GetValues<uint8_t>();
+  auto num_channels = format_->InputNeededPerElement();
+  uint32_t channel_id = 0;
+
+  for (size_t i = 0; i < ElementCount(); ++i) {
+    for (const auto& seg : format_->GetSegments()) {
+      if (seg.IsPadding()) {
+        buf_ptr += seg.PaddingBytes();
+        continue;
+      }
+      if (channel_id == channel) {
+        assert(type::Type::IsUint8(seg.GetFormatMode(), seg.GetNumBits()));
+        const auto bin = *reinterpret_cast<const uint8_t*>(buf_ptr);
+        bins[bin]++;
+      }
+      buf_ptr += seg.SizeInBytes();
+      channel_id = (channel_id + 1) % num_channels;
+    }
+  }
+
+  return bins;
+}
+
+Result Buffer::CompareHistogramEMD(Buffer* buffer, float tolerance) const {
+  auto result = CheckCompability(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  const int num_bins = 256;
+  auto num_channels = format_->InputNeededPerElement();
+  for (auto segment : format_->GetSegments()) {
+    if (!type::Type::IsUint8(segment.GetFormatMode(), segment.GetNumBits()) ||
+        num_channels != 4)
+      return Result(
+          "EMD comparison only supports 8bit unorm format with four channels.");
+  }
+
+  std::vector<std::vector<int64_t>> histogram1, histogram2;
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    histogram1.push_back(GetHistogramForChannel(c, num_bins));
+    histogram2.push_back(buffer->GetHistogramForChannel(c, num_bins));
+  }
+
+  // Earth movers's distance: Calculate the minimal cost of moving "earth" to
+  // transform the first histogram into the second, where each bin of the
+  // histogram can be thought of as a column of units of earth. The cost is the
+  // amount of earth moved times the distance carried (the distance is the
+  // number of adjacent bins over which the earth is carried). Calculate this
+  // using the cumulative difference of the bins, which works as long as both
+  // histograms have the same amount of earth. Sum the absolute values of the
+  // cumulative difference to get the final cost of how much (and how far) the
+  // earth was moved.
+  double max_emd = 0;
+
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    uint64_t diff_total = 0;
+    int64_t diff_accum = 0;
+
+    for (size_t i = 0; i < histogram1[c].size(); ++i) {
+      diff_accum += histogram1[c][i] - histogram2[c][i];
+      diff_total += std::abs(diff_accum);
+    }
+    // Normalize to range 0..1
+    double emd = static_cast<double>(diff_total) / (num_bins * element_count_);
+    max_emd = std::max(max_emd, emd);
+  }
+
+  if (max_emd > static_cast<double>(tolerance)) {
+    return Result("Histogram EMD value of " + std::to_string(max_emd) +
+                  " is greater than tolerance of " + std::to_string(tolerance));
+  }
+
+  return {};
```

Comment:
Nit: doesn't matter for this PR, but when adding to the standalone `image_diff` tool, I guess you might want to move the code that yields `max_emd` into another (public?) function so that `image_diff` can get and print the value, even if the tolerance was not exceeded.